src/systems/filecoin_vm/sysactors/reward_actor.go

```diff
@@ -74,12 +89,56 @@ func (a *RewardActorCode_I) Constructor(rt vmr.Runtime) InvocOutput {
 	panic("TODO")
 }
 
-func (a *RewardActorCode_I) AwardBlockReward(rt vmr.Runtime, miner addr.Address, penalty actor.TokenAmount) {
+func (a *RewardActorCode_I) AwardBlockReward(
+	rt vmr.Runtime,
+	miner addr.Address,
+	penalty actor.TokenAmount,
+	minerActivePower block.StoragePower,
+	minerInactivePower block.StoragePower,
+	currPledge actor.TokenAmount,
+) {
 	rt.ValidateImmediateCallerIs(addr.SystemActorAddr)
-	// block reward function should live here
-	// handle penalty greater than reward
-	// put Reward into RewardMap
-	panic("TODO")
+
+	inds := rt.CurrIndices()
+	pledgeReq := inds.BlockReward_PledgeCollateralReq(minerActivePower, minerInactivePower, currPledge)
+	currReward := inds.BlockReward_GetCurrRewardForMiner(minerActivePower, currPledge)
+	underPledge := math.Max(float64(actor.TokenAmount(0)), float64(pledgeReq-currPledge)) // 0 if over collateralized
+	rewardToGarnish := math.Min(float64(currReward), float64(underPledge))
+
+	util.TODO()
+	// handle penalty here
+	// also handle penalty greater than reward
+	actualReward := currReward - actor.TokenAmount(rewardToGarnish)
+	if rewardToGarnish > 0 {
+		// Send fund to SPA for collateral
+		rt.Send(
+			addr.StoragePowerActorAddr,
+			ai.Method_StoragePowerActor_AddBalance,
+			[]util.Serialization{
+				addr.Serialize_Address(miner),
+			},
+			actor.TokenAmount(rewardToGarnish),
+		)
+	}
+
+	h, st := a.State(rt)
+	if actualReward > 0 {
+		// put Reward into RewardMap
+		newReward := &Reward_I{
+			StartEpoch_:      rt.CurrEpoch(),
+			EndEpoch_:        rt.CurrEpoch(),
+			Value_:           actualReward,
+			AmountWithdrawn_: actor.TokenAmount(0),
+			VestingFunction_: VestingFunction_None,
```

Comment:
Noting that if we actually decide to have no vesting, we should skip all this complexity and just pay the reward to the owner immediately. This is very heavy on chain.

Reply:
Definitely